src/metabase/models/serialization/base.clj

```diff
@@ -472,3 +498,136 @@
   (if (entity-id? id-str)
     (db/select-one model :entity_id id-str)
     (find-by-identity-hash model id-str)))
+
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                                Storage                                                         |
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; These storage multimethods take a second argument known as the context. This is a good place for a particular
+;;; storage implementation to include some precomputed information, or options.
+;;; In particular, it should include a table of collection IDs to path fragments that is precomputed by the host.
+;;; [[storage-base-context]] computes that, since many things go in a collections tree.
+(def ^:private max-label-length 100)
+
+(defn- truncate-label [s]
+  (if (> (count s) max-label-length)
+      (subs s 0 max-label-length)
+      s))
+
+(defn- lower-plural [s]
+  (-> s u/lower-case-en (str "s")))
+
+(defn storage-leaf-file-name
+  "Captures the common pattern for leaf file names as `entityID_label`."
+  ([id]       (str id))
+  ([id label] (if (nil? label)
+                (storage-leaf-file-name id)
+                (str id "_" (truncate-label label)))))
+
+(defn storage-default-collection-path
+  "Implements the most common structure for [[storage-path]] - `collections/c1/c2/c3/models/entityid_slug.ext`"
+  [entity {:keys [collections]}]
+  (let [{:keys [model id label]} (-> entity serdes-path last)
+        result
+    (concat ["collections"]
+            (get collections (:collection_id entity)) ;; This can be nil, but that's fine - that's the root collection.
+            [(lower-plural model) (storage-leaf-file-name id label)])]
+    #_(clojure.pprint/pprint {:entity entity :path result})
+    result))
+
+(defmulti storage-path
+  "Computes the complete storage path for a given entity.
+  `(storage-path entity ctx)`
+  Dispatches on the model name, eg. \"Dashboard\".
+
+  Returns a list of strings giving the path, with the final entry being the file name with no extension.
+
+  The default implementation works for entities which are:
+  - Part of the regular (not snippet) collections tree, per a :collection_id field; and
+  - Stored as `foos/1234abc_slug.extension` underneath their collection
+  eg. Cards, Dashboards, Timelines
+
+  The default logic is captured by [[storage-default-collection-path]] so it can be reused."
+  (fn [entity _] (ingested-model entity)))
```

Comment:
Can you add `{:arglists '([...])` metadata here? So it pops up in eldoc or equivalent when you're trying to write a new usage of it. Otherwise you have to read the entire docstring to try to figure out how many arguments this takes and what they are 

```suggestion
  {:arglists '([entity ctx])}
  (fn [entity _] (ingested-model entity)))
```

Reply:
Done.